File: LeetCode/python-solutions/y2023/m7/v30/detect_cycle.py
# -*- coding: utf-8 -*-
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return None

        common_node: Optional[ListNode] = self.getCommonNode(head)
        if common_node is None:
            return None
        current_node: Optional[ListNode] = head
        while current_node != common_node:
            current_node, common_node = current_node.next, common_node.next

        return current_node

    @classmethod
    def getCommonNode(cls, head: ListNode) -> Optional[ListNode]:
        slow, fast = head, head

        while not (fast is None or fast.next is None):
            slow, fast = slow.next, fast.next.next

            if slow == fast:
                return slow

        return None

File: LeetCode/go-solutions/y2023/m7/v30/detect_cycle.go
package main

// ListNode Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	// 空链表或单节点不可能有环，直接返回
	if head == nil || head.Next == nil || head.Next.Next == nil {
		return nil
	}

	// 假设相遇点距入环点的距离为 b，环总长为 b + c，fast 经过 n 圈后追上
	// slow -> a + b
	// fast -> a + n(b + c)
	// n = 1 -> fast 和 slow 在环中最远相距 b + c - 1，假设 slow 静止，fast 速度为 1，肯定可以在 b + c - 1 秒内追上 slow
	// 此时 slow 未走完一圈
	// a + b = a + n(b + c) -> a = (n - 1)(b + c) + c -> a = c
	commonNode := getCommonNode(head)
	if commonNode == nil {
		return nil
	}

	currentNode := head
	for commonNode != currentNode {
		commonNode = commonNode.Next
		currentNode = currentNode.Next
	}

	return currentNode
}

func getCommonNode(head *ListNode) *ListNode {
	slow := head
	fast := head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return slow
		}
	}

	return nil
}
